cdc/processor/sinkmanager/worker_impl.go

```diff
@@ -221,3 +240,115 @@ func (w *workerImpl) advanceTableSink(t *tableSinkTask, commitTs model.Ts, size
 	w.memQuota.record(t.tableID, resolvedTs, size)
 	return t.tableSink.updateResolvedTs(resolvedTs)
 }
+
+type redoWorkerImpl struct {
+	changefeedID   model.ChangeFeedID
+	sortEngine     sorter.EventSortEngine
+	memQuota       *memQuota
+	redoManager    redo.LogManager
+	eventCache     *redoEventCache
+	splitTxn       bool
+	enableOldValue bool
+}
+
+func newRedoWorker(
+	changefeedID model.ChangeFeedID,
+	sortEngine sorter.EventSortEngine,
+	quota *memQuota,
+	redoManager redo.LogManager,
+	eventCache *redoEventCache,
+	splitTxn bool,
+	enableOldValue bool,
+) redoWorker {
+	return &redoWorkerImpl{
+		changefeedID:   changefeedID,
+		sortEngine:     sortEngine,
+		memQuota:       quota,
+		redoManager:    redoManager,
+		eventCache:     eventCache,
+		splitTxn:       splitTxn,
+		enableOldValue: enableOldValue,
+	}
+}
+
+func (w *redoWorkerImpl) handleTasks(ctx context.Context, taskChan <-chan *redoTask) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case task := <-taskChan:
+			rows := make([]*model.RowChangedEvent, 0, 1024)
+			batchSize := uint64(0)
+			cache := w.eventCache.getAppender(task.tableID)
+			cachedSize := uint64(0)
+			memAllocated := true
+
+			var lastPos sorter.Position
+			var maybeEmitBatchEvents = func(allFinished, txnFinished bool) error {
+				if batchSize == 0 || (!allFinished && batchSize < requestMemSize) {
+					return nil
+				}
+
+				releaseMem := func() { w.memQuota.refund(batchSize - cachedSize) }
+				err := w.redoManager.EmitRowChangedEvents(ctx, task.tableID, releaseMem, rows...)
+				if err != nil {
+					return errors.Trace(err)
+				}
+				if lastPos.Valid() {
+					err = w.redoManager.UpdateResolvedTs(ctx, task.tableID, lastPos.CommitTs)
+					if err != nil {
+						return errors.Trace(err)
+					}
+				}
+
+				rows = rows[0:]
+				if cap(rows) > 1024 {
+					rows = make([]*model.RowChangedEvent, 0, 1024)
+				}
+				batchSize = 0
+				cachedSize = 0
+
+				if !allFinished {
+					if !txnFinished {
+						w.memQuota.forceAcquire(requestMemSize)
+					} else {
+						memAllocated = w.memQuota.tryAcquire(requestMemSize)
+					}
+				}
+				return nil
+			}
+
+			// lowerBound and upperBound are both closed intervals.
+			iter := w.sortEngine.FetchByTable(task.tableID, task.lowerBound, task.getUpperBound())
+			defer iter.Close()
```

Comment:
Please do not use `defer` in for loop. I guess it will not be executed because this func never stops.